Add documentation for Vara Bridge (#220)
gear-foundation/vara-wiki: docs/bridge/architecture.md

## Overview
This page provides a technical overview of the Vara Bridge architecture, without covering every component in exhaustive detail. The goal is to give readers a high-level understanding of how the Bridge operates. For complete technical information, please refer to the [Gear Bridges repository](https://github.com/gear-tech/gear-bridges).

Vara Bridge uses Plonky2 circuits to implement cryptographic primitives, forming a secure foundation for cross-chain operations. Plonky2 is a zero-knowledge (ZK) proof system that provides efficient and scalable proofs. By using these circuits, Vara Bridge preserves transaction integrity and greatly reduces the possibility of manipulation. The proof mechanism also guarantees that specific pallet states are finalized and included in the Vara Network before those states are relayed to Ethereum. Verified within a Gnark circuit, these proofs help maintain trust and accuracy as data moves across chains.

## Bridge Components
The Vara Bridge consists of several distinct parts, including on-chain components on both Vara and Ethereum, as well as off-chain components for securely relaying message proofs.

### Components on Vara Network

> - **VFT programs**: Deployments of the VFT token standard, capable of transferring, burning, and minting VFT tokens. Each unique token has its own VFT standard deployment on Vara.

Reviewer: its own VFT standard? sounds like they are based on different standards. They all are implemented via the same standard. And the link would be useful: /docs/build/standards/vft

> - **VFT Manager**: A single application that receives VFT tokens from users, burns them, and emits a message to the `pallet-gear-eth-bridge` Built-in actor. This message indicates which token is being bridged, the token amount, and the intended Ethereum recipient.

Reviewer 2: > A single application

Maybe program?

- **Bridging Built-in Actor**: A new Built-in Actor serving as the entry point for the generic bridging protocol. It takes messages from any actor on the Vara Network and forwards them to the Bridging pallet.

> - **Bridging Pallet**: Accepts messages from the Bridge Built-in Actor and stores them in specialized data (a binary Merkle trie).

Reviewer 2: ```suggestion
- **Bridging Pallet**: Accepts messages from the Bridge Built-in Actor and stores them in specialized data structure (a binary Merkle trie).
```

> - **Ethereum-Event-Client**: A program on Vara that can verify whether a specific event was included in a particular block. It consults the `checkpoint-light-client` to confirm the validity of that block.

Reviewer 2: Maybe clarify here that these events come from the smart-contracts on ethereum?

- **Historical-Proxy**: A program on Vara that maintains historical `ethereum-event-client` program addresses and routes requests to the `ethereum-event-client` instance responsible for processing the requested transaction.

### Components on Ethereum

> - **ERC20Treasury**: Implements the `IMessageQueueReceiver` interface and functions as a vault for ERC-20 tokens, permitting users to deposit tokens and manage withdrawals. It processes messages relayed by the `MessageQueue` and accepts only properly formatted and authorized messages.

Reviewer 2: ```suggestion
- **ERC20Manager**: Implements the `IMessageQueueReceiver` interface and functions as a vault for ERC-20 tokens, permitting users to deposit tokens and manage withdrawals. It processes messages relayed by the `MessageQueue` and accepts only properly formatted and authorized messages.
```

> - **MessageQueue**: Verifies proofs of message inclusion in a Merkle trie, whose root is stored in the `Relayer`. It ensures each message is processed exactly once by comparing the computed Merkle root with the root kept in the `Relayer`. After successful verification, it calls the target `IMessageQueueReceiver` contract (e.g., `ERC20Treasury`).

Reviewer 2: ```suggestion
- **MessageQueue**: Verifies proofs of message inclusion in a Merkle trie, whose root is stored in the `Relayer`. It ensures each message is processed exactly once by storing nonces for all of the processed messages. After successful verification, it calls the target `IMessageQueueReceiver` contract (e.g., `ERC20Treasury`).
```

> - **Relayer**: Receives a Merkle trie root proof from the Prover and submits it to the `Verifier` for validation using `PlonkVerifier`. Once the proof is confirmed, the `Relayer` records the validated Merkle trie root and block number, so the `MessageQueue` can authenticate messages.

Reviewer 2: The naming for `Relayer` is a little legacy, so maybe it'd be worth to mention that `Relayer` means contract on ethereum and not an actual relayer

> - **Verifier**: Uses a `PlonkVerifier` to check zero-knowledge proofs demonstrating that a Merkle trie root belongs to a finalized Vara block. By verifying the Merkle trie root’s integrity, it ensures the data that arrives at the `Relayer` is genuine, facilitating the rest of the bridging process.

Reviewer 2: ```suggestion
- **Verifier**: Uses a `PlonkVerifier` to check gnark proofs demonstrating that a Merkle trie root belongs to a finalized Vara block. By verifying the Merkle trie root’s integrity, it ensures the data that arrives at the `Relayer` is genuine, facilitating the rest of the bridging process.
```

> - **Verifier**: Uses a `PlonkVerifier` to check zero-knowledge proofs demonstrating that a Merkle trie root belongs to a finalized Vara block. By verifying the Merkle trie root’s integrity, it ensures the data that arrives at the `Relayer` is genuine, facilitating the rest of the bridging process.

Reviewer 2: It doesn't verify integrity of the merkle roots, it just checks that they're the same as in the proof

Author: Thank you, @mertwole. Please check again.

### Off-Chain Components
In addition to the on-chain components, there are relayer nodes responsible for transferring data between Vara and Ethereum:

- **Vara→Ethereum Protocol relayer**: Reads the Vara state, creates ZK proofs, and sends them to Ethereum. It produces two kinds of ZK proofs:  
  - Proof of authority set changes  
  - Proof of Merkle trie root inclusion in the `pallet-gear-eth-bridge` storage

- **Ethereum→Vara Protocol relayer**: Tracks `sync committee` changes and the blocks they sign on Ethereum. It then updates the `checkpoint-light-client` state on Vara.

- **Ethereum→Vara Token relayer** and **Vara→Ethereum Token relayer**: Monitor events from `bridging-payment` services and handle cross-chain actions to ensure messages are delivered correctly.

For transfers from Ethereum to the Vara Network, there is also a lazy Ethereum light client:

> - **Checkpoint-Light-Client**: A lazy Ethereum light client that keeps track of the `sync committee` validator set and confirms block headers using it.

Reviewer 2: It's not an off-chain service, it's a program on VARA
